logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java

```diff
@@ -0,0 +1,276 @@
+        if (this.elementCount <= 0) {
+            this.minSeqNum = 0;
+        }
+    }
+
+    // we don't have different versions yet so simply check if the version is VERSION_ONE for basic integrity check
+    // and if an unexpected version byte is read throw PageIOInvalidVersionException
+    private void validateVersion(byte version) throws PageIOInvalidVersionException {
+        if (version != VERSION_ONE) {
+            throw new PageIOInvalidVersionException(String.format("Expected page version=%d but found version=%d", VERSION_ONE, version));
+        }
+    }
+
+    // read and vadidate next element at page head
+    // @param verifyChecksum if true the actual element data will be read + checksumed and compared to written checksum
+    private void readNextElement(long expectedSeqNum, boolean verifyChecksum) throws PageIOInvalidElementException {
+        // if there is no room for the seqNum and length bytes stop here
+        if (this.head + SEQNUM_SIZE + LENGTH_SIZE > capacity) { throw new PageIOInvalidElementException("cannot read seqNum and length bytes past buffer capacity"); }
+
+        int elementOffset = this.head;
+        ByteBuffer buffer = getBuffer();
+
+        long seqNum = buffer.getLong();
+        this.head += SEQNUM_SIZE;
+
+        if (seqNum != expectedSeqNum) { throw new PageIOInvalidElementException(String.format("Element seqNum %d is expected to be %d", seqNum, expectedSeqNum)); }
+
+        int length = buffer.getInt();
+        this.head += LENGTH_SIZE;
+
+        // length must be > 0
+        if (length <= 0) { throw new PageIOInvalidElementException("Element invalid length"); }
+
+        // if there is no room for the proposed data length and checksum just stop here
+        if (this.head + length + CHECKSUM_SIZE > capacity) { throw new PageIOInvalidElementException("cannot read element payload and checksum past buffer capacity"); }
+
+        if (verifyChecksum) {
+            // read data and compute checksum;
+            byte[] readBytes = new byte[length];
+            buffer.get(readBytes);
+            int checksum = buffer.getInt();
+            int computedChecksum = checksum(readBytes);
+            if (computedChecksum != checksum) { throw new PageIOInvalidElementException("Element invalid checksum"); }
+        }
+
+        // at this point we recovered a valid element
+        this.offsetMap.add(elementOffset);
+        this.head += length + CHECKSUM_SIZE;
+
+        buffer.position(head);
+    }
+
+    @Override
+    public void create() throws IOException {
+        getBuffer().position(0);
+        getBuffer().put(VERSION_ONE);
+        this.head = 1;
+        this.minSeqNum = 0L;
+        this.elementCount = 0;
+    }
+
+    @Override
+    public void write(byte[] bytes, long seqNum) throws IOException {
+        write(bytes, seqNum, bytes.length, checksum(bytes));
+    }
+
+    protected int write(byte[] bytes, long seqNum, int length, int checksum) {
```

Comment:
yes, because it allows forcing each value independently of the actual bytes content. note that this is not a public method, and it is used by tests to inject incorrect values.